File: ios/TokenTracker/Features/Dashboard/DashboardState.swift
import Foundation

struct ProviderReport: Equatable {
    let provider: ProviderKind
    let orgName: String
    let report: MTDCost
}

enum DashboardState: Equatable {
    case needsCredentials
    case idle
    case loading
    case loaded(report: MTDCost, orgName: String)
    case failed(message: String)

    var isLoaded: Bool {
        if case .loaded = self { return true }
        return false
    }

    var orgName: String? {
        if case .loaded(_, let name) = self { return name }
        return nil
    }
}


extension Dictionary where Key == ProviderKind, Value == (report: MTDCost, orgName: String) {
    var providerReports: [ProviderReport] {
        sorted { $0.key.rawValue < $1.key.rawValue }
            .map { provider, cached in
                ProviderReport(provider: provider, orgName: cached.orgName, report: cached.report)
            }
    }
}

extension Array {
    func ifEmpty(_ fallback: [Element]) -> [Element] { isEmpty ? fallback : self }
}

File: android/app/src/main/kotlin/studio/maximumimpact/tokencounter/features/dashboard/DashboardState.kt
package studio.maximumimpact.tokencounter.features.dashboard

import studio.maximumimpact.tokencounter.core.MtdCost
import studio.maximumimpact.tokencounter.providers.ProviderKind

/** One provider's contribution to a loaded dashboard. */
data class ProviderReport(
    val provider: ProviderKind,
    val orgName: String,
    val report: MtdCost
)

/**
 * Mirror of the iOS `DashboardView.DashboardState` enum. The whole app is a
 * single state machine driven off this value; [studio.maximumimpact.tokencounter.TokenCounterApp]
 * switches the rendered screen on the case.
 */
sealed interface DashboardState {
    data object Loading : DashboardState
    data object NeedsCredentials : DashboardState
    data class Loaded(
        val orgName: String,
        val report: MtdCost,
        val providerReports: List<ProviderReport> = emptyList(),
        val selectedProvider: ProviderKind? = null
    ) : DashboardState
    data class Failed(val message: String) : DashboardState
}
